### PR TITLE
VK provider fix

### PR DESCRIPTION
**Pull request by alexTayanovsky**
## Reasoning 💡

VK provider is broken (see https://github.com/nextauthjs/next-auth/discussions/3587, https://github.com/nextauthjs/next-auth/issues/2524#issuecomment-967222842), so this PR aims to fix it, and adds Typescript support.

## Checklist 🧢

- [x] Documentation
- [x] Tests
- [x] Ready to be merged

### CODE REVIEW

File: src/providers/vk.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    photo_100: string
+    can_access_closed: boolean
+    is_closed: boolean
+  }>
+}
+
+export default function VK<
+  P extends Record<string, any> = VkProfile
+>(options: OAuthUserConfig<P>): OAuthConfig<P> {
+  const apiVersion = "5.126" // https://vk.com/dev/versions
```
Comment (balazsorban44) [inline src/providers/vk.ts:17]:
```suggestion
  const apiVersion = "5.131" // https://vk.com/dev/versions
```

I tested this, let's update the version.

File: src/providers/vk.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+export interface VkProfile {
+  response: Array<{
+    id: number
+    first_name: string
+    last_name: string
+    photo_100: string
+    can_access_closed: boolean
+    is_closed: boolean
+  }>
```
Comment (balazsorban44) [inline src/providers/vk.ts:11]:
```suggestion
  /** [Additional Fields 1](https://vk.com/dev/fields) | [Additional Fields 2](https://vk.com/dev/fields_2) */
  response: Array<{
    id: number
    first_name: string
    last_name: string
    photo_100: string
    can_access_closed: boolean
    is_closed: boolean
  }>
```

I found a bunch of extra fields, let's add it as a JSDoc.

Feel free to add all of those to the interface as well, for easier discovery for the user.

**Review by balazsorban44 — CHANGES_REQUESTED** [review]
Thank you. I actually created a VK app on my own and added a few suggestions to this PR.

Looks like there is also a merge conflict, please fix it as well!

**Comment by AtomBaytovich** [comment]
Hi! I ran into the same problem today, also a bug 'invalid client (client ID not defined)'.

**Comment by alexTayanovsky** [comment]
@balazsorban44 Thank you very much for assistance! I've addressed your review comments & suggestions

**Comment by eolme** [comment]
Any updates here?

**Comment by sergeyshaykhullin** [comment]
@balazsorban44 I tested this fix, it is working, but was unable to save authorized user inside db

```
[next-auth][error][OAUTH_CALLBACK_HANDLER_ERROR]
https://next-auth.js.org/errors#oauth_callback_handler_error
Invalid `p.account.create()` invocation in
C:\REDACTED\app\node_modules\@next-auth\prisma-adapter\dist\index.js:19:42

  16 },
  17 updateUser: ({ id, ...data }) => p.user.update({ where: { id }, data }),
  18 deleteUser: (id) => p.user.delete({ where: { id } }),
→ 19 linkAccount: (data) => p.account.create({
       data: {
         provider: 'vk',
         type: 'oauth',
         providerAccountId: 'REDACTED',
         access_token: 'REDACTED',
         expires_at: 1651408254,
         user_id: REDACTED,
         ~~~~~~~
         email: 'REDACTED',
         ~~~~~
         userId: 'REDACTED'
       }
     })

Unknown arg `user_id` in data.user_id for type AccountUncheckedCreateInput. Did you mean `userId`? Available args:
type AccountUncheckedCreateInput {
  id?: String
  userId: String
  type: String
  provider: String
  providerAccountId: String
  refresh_token?: String | Null
  access_token?: String | Null
  expires_at?: Int | Null
  token_type?: String | Null
  scope?: String | Null
  id_token?: String | Null
  session_state?: String | Null
  oauth_token_secret?: String | Null
  oauth_token?: String | Null
}
Unknown arg `email` in data.email for type AccountUncheckedCreateInput. Did you mean `id`? Available args:
type AccountUncheckedCreateInput {
  id?: String
  userId: String
  type: String
  provider: String
  providerAccountId: String
  refresh_token?: String | Null
  access_token?: String | Null
  expires_at?: Int | Null
  token_type?: String | Null
  scope?: String | Null
  id_token?: String | Null
  session_state?: String | Null
  oauth_token_secret?: String | Null
  oauth_token?: String | Null
}

 Error:
Invalid `p.account.create()` invocation in
C:\REDACTED\app\node_modules\@next-auth\prisma-adapter\dist\index.js:19:42

  16 },
  17 updateUser: ({ id, ...data }) => p.user.update({ where: { id }, data }),
  18 deleteUser: (id) => p.user.delete({ where: { id } }),
→ 19 linkAccount: (data) => p.account.create({
       data: {
         provider: 'vk',
         type: 'oauth',
         providerAccountId: 'REDACTED',
         access_token: 'REDACTED',
         expires_at: 1651408254,
         user_id: REDACTED,
         ~~~~~~~
         email: 'REDACTED',
         ~~~~~
         userId: 'REDACTED'
       }
     })

Unknown arg `user_id` in data.user_id for type AccountUncheckedCreateInput. Did you mean `userId`? Available args:
type AccountUncheckedCreateInput {
  id?: String
  userId: String
  type: String
  provider: String
  providerAccountId: String
  refresh_token?: String | Null
  access_token?: String | Null
  expires_at?: Int | Null
  token_type?: String | Null
  scope?: String | Null
  id_token?: String | Null
  session_state?: String | Null
  oauth_token_secret?: String | Null
  oauth_token?: String | Null
}
Unknown arg `email` in data.email for type AccountUncheckedCreateInput. Did you mean `id`? Available args:
type AccountUncheckedCreateInput {
  id?: String
  userId: String
  type: String
  provider: String
  providerAccountId: String
  refresh_token?: String | Null
  access_token?: String | Null
  expires_at?: Int | Null
  token_type?: String | Null
  scope?: String | Null
  id_token?: String | Null
  session_state?: String | Null
  oauth_token_secret?: String | Null
  oauth_token?: String | Null
}


    at Object.validate (C:\REDACTED\app\node_modules\@prisma\client\runtime\index.js:44425:20)
    at PrismaClient._executeRequest (C:\REDACTED\app\node_modules\@prisma\client\runtime\index.js:46509:17)
    at consumer (C:\REDACTED\app\node_modules\@prisma\client\runtime\index.js:46453:23)
    at C:\REDACTED\app\node_modules\@prisma\client\runtime\index.js:46457:76
    at runInChildSpan (C:\REDACTED\app\node_modules\@prisma\client\runtime\index.js:45702:12)
    at C:\REDACTED\app\node_modules\@prisma\client\runtime\index.js:46457:20
    at AsyncResource.runInAsyncScope (node:async_hooks:201:9)
    at PrismaClient._request (C:\REDACTED\app\node_modules\@prisma\client\runtime\index.js:46456:86)
    at C:\REDACTED\app\node_modules\@prisma\client\runtime\index.js:43088:25
    at _callback (C:\REDACTED\app\node_modules\@prisma\client\runtime\index.js:42852:52) {
  name: 'LinkAccountError',
  code: undefined
}
```

**Comment by sergeyshaykhullin** [comment]
Seems like Mailru provider also broken

```
[next-auth][error][OAUTH_PARSE_PROFILE_ERROR]
23
https://next-auth.js.org/errors#oauth_parse_profile_error Cannot read properties of undefined (reading 'toString') {
22
  error: {
21
    message: "Cannot read properties of undefined (reading 'toString')",
20
    stack: "TypeError: Cannot read properties of undefined (reading 'toString')\n" +
19
      '    at getProfile (/app/node_modules/next-auth/core/lib/oauth/callback.js:176:39)\n' +
18
      '    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n' +
17
      '    at async oAuthCallback (/app/node_modules/next-auth/core/lib/oauth/callback.js:139:27)\n' +
16
      '    at async Object.callback (/app/node_modules/next-auth/core/routes/callback.js:50:11)\n' +
15
      '    at async NextAuthHandler (/app/node_modules/next-auth/core/index.js:139:28)\n' +
14
      '    at async NextAuthNextHandler (/app/node_modules/next-auth/next/index.js:21:19)\n' +
13
      '    at async /app/node_modules/next-auth/next/index.js:57:32\n' +
12
      '    at async Object.apiResolver (/app/node_modules/next/dist/server/api-utils/node.js:184:9)\n' +
11
      '    at async NextNodeServer.runApi (/app/node_modules/next/dist/server/next-server.js:397:9)\n' +
10
      '    at async Object.fn (/app/node_modules/next/dist/server/base-server.js:473:37)',
9
    name: 'TypeError'
8
  },
7
  OAuthProfile: {
6
    error: 'invalid request',
5
    error_code: 2,
4
    error_description: 'missing request parameter: access_token'
3
  },
2
  message: "Cannot read properties of undefined (reading 'toString')"
1
}
```

**Comment by ndom91** [comment]
> Seems like Mailru provider is also broken
> 
> ```
> [next-auth][error][OAUTH_PARSE_PROFILE_ERROR]
> 23
> https://next-auth.js.org/errors#oauth_parse_profile_error Cannot read properties of undefined (reading 'toString') {
> 22
>   error: {
> 21
>     message: "Cannot read properties of undefined (reading 'toString')",
> 20
>     stack: "TypeError: Cannot read properties of undefined (reading 'toString')\n" +
> 19
>       '    at getProfile (/app/node_modules/next-auth/core/lib/oauth/callback.js:176:39)\n' +
> 18
>       '    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n' +
> 17
>       '    at async oAuthCallback (/app/node_modules/next-auth/core/lib/oauth/callback.js:139:27)\n' +
> 16
>       '    at async Object.callback (/app/node_modules/next-auth/core/routes/callback.js:50:11)\n' +
> 15
>       '    at async NextAuthHandler (/app/node_modules/next-auth/core/index.js:139:28)\n' +
> 14
>       '    at async NextAuthNextHandler (/app/node_modules/next-auth/next/index.js:21:19)\n' +
> 13
>       '    at async /app/node_modules/next-auth/next/index.js:57:32\n' +
> 12
>       '    at async Object.apiResolver (/app/node_modules/next/dist/server/api-utils/node.js:184:9)\n' +
> 11
>       '    at async NextNodeServer.runApi (/app/node_modules/next/dist/server/next-server.js:397:9)\n' +
> 10
>       '    at async Object.fn (/app/node_modules/next/dist/server/base-server.js:473:37)',
> 9
>     name: 'TypeError'
> 8
>   },
> 7
>   OAuthProfile: {
> 6
>     error: 'invalid request',
> 5
>     error_code: 2,
> 4
>     error_description: 'missing request parameter: access_token'
> 3
>   },
> 2
>   message: "Cannot read properties of undefined (reading 'toString')"
> 1
> }
> ```

Can you open a separate issue for `mail.ru` provider?

**Comment by ndom91** [comment]
> @balazsorban44 I tested this fix, it is working, but was unable to save authorized user inside db
> 
> <details>
> <summary>Log Output</summary>
> 
> ```
> [next-auth][error][OAUTH_CALLBACK_HANDLER_ERROR]
> https://next-auth.js.org/errors#oauth_callback_handler_error
> Invalid `p.account.create()` invocation in
> C:\REDACTED\app\node_modules\@next-auth\prisma-adapter\dist\index.js:19:42
> 
>   16 },
>   17 updateUser: ({ id, ...data }) => p.user.update({ where: { id }, data }),
>   18 deleteUser: (id) => p.user.delete({ where: { id } }),
> → 19 linkAccount: (data) => p.account.create({
>        data: {
>          provider: 'vk',
>          type: 'oauth',
>          providerAccountId: 'REDACTED',
>          access_token: 'REDACTED',
>          expires_at: 1651408254,
>          user_id: REDACTED,
>          ~~~~~~~
>          email: 'REDACTED',
>          ~~~~~
>          userId: 'REDACTED'
>        }
>      })
> 
> Unknown arg `user_id` in data.user_id for type AccountUncheckedCreateInput. Did you mean `userId`? Available args:
> type AccountUncheckedCreateInput {
>   id?: String
>   userId: String
>   type: String
>   provider: String
>   providerAccountId: String
>   refresh_token?: String | Null
>   access_token?: String | Null
>   expires_at?: Int | Null
>   token_type?: String | Null
>   scope?: String | Null
>   id_token?: String | Null
>   session_state?: String | Null
>   oauth_token_secret?: String | Null
>   oauth_token?: String | Null
> }
> Unknown arg `email` in data.email for type AccountUncheckedCreateInput. Did you mean `id`? Available args:
> type AccountUncheckedCreateInput {
>   id?: String
>   userId: String
>   type: String
>   provider: String
>   providerAccountId: String
>   refresh_token?: String | Null
>   access_token?: String | Null
>   expires_at?: Int | Null
>   token_type?: String | Null
>   scope?: String | Null
>   id_token?: String | Null
>   session_state?: String | Null
>   oauth_token_secret?: String | Null
>   oauth_token?: String | Null
> }
> 
>  Error:
> Invalid `p.account.create()` invocation in
> C:\REDACTED\app\node_modules\@next-auth\prisma-adapter\dist\index.js:19:42
> 
>   16 },
>   17 updateUser: ({ id, ...data }) => p.user.update({ where: { id }, data }),
>   18 deleteUser: (id) => p.user.delete({ where: { id } }),
> → 19 linkAccount: (data) => p.account.create({
>        data: {
>          provider: 'vk',
>          type: 'oauth',
>          providerAccountId: 'REDACTED',
>          access_token: 'REDACTED',
>          expires_at: 1651408254,
>          user_id: REDACTED,
>          ~~~~~~~
>          email: 'REDACTED',
>          ~~~~~
>          userId: 'REDACTED'
>        }
>      })
> 
> Unknown arg `user_id` in data.user_id for type AccountUncheckedCreateInput. Did you mean `userId`? Available args:
> type AccountUncheckedCreateInput {
>   id?: String
>   userId: String
>   type: String
>   provider: String
>   providerAccountId: String
>   refresh_token?: String | Null
>   access_token?: String | Null
>   expires_at?: Int | Null
>   token_type?: String | Null
>   scope?: String | Null
>   id_token?: String | Null
>   session_state?: String | Null
>   oauth_token_secret?: String | Null
>   oauth_token?: String | Null
> }
> Unknown arg `email` in data.email for type AccountUncheckedCreateInput. Did you mean `id`? Available args:
> type AccountUncheckedCreateInput {
>   id?: String
>   userId: String
>   type: String
>   provider: String
>   providerAccountId: String
>   refresh_token?: String | Null
>   access_token?: String | Null
>   expires_at?: Int | Null
>   token_type?: String | Null
>   scope?: String | Null
>   id_token?: String | Null
>   session_state?: String | Null
>   oauth_token_secret?: String | Null
>   oauth_token?: String | Null
> }
> 
> 
>     at Object.validate (C:\REDACTED\app\node_modules\@prisma\client\runtime\index.js:44425:20)
>     at PrismaClient._executeRequest (C:\REDACTED\app\node_modules\@prisma\client\runtime\index.js:46509:17)
>     at consumer (C:\REDACTED\app\node_modules\@prisma\client\runtime\index.js:46453:23)
>     at C:\REDACTED\app\node_modules\@prisma\client\runtime\index.js:46457:76
>     at runInChildSpan (C:\REDACTED\app\node_modules\@prisma\client\runtime\index.js:45702:12)
>     at C:\REDACTED\app\node_modules\@prisma\client\runtime\index.js:46457:20
>     at AsyncResource.runInAsyncScope (node:async_hooks:201:9)
>     at PrismaClient._request (C:\REDACTED\app\node_modules\@prisma\client\runtime\index.js:46456:86)
>     at C:\REDACTED\app\node_modules\@prisma\client\runtime\index.js:43088:25
>     at _callback (C:\REDACTED\app\node_modules\@prisma\client\runtime\index.js:42852:52) {
>   name: 'LinkAccountError',
>   code: undefined
> }
> ```
>
> </details>

So theres two errors here according to your output:

1. `account` model / table doesnt have a field called `email`
2. The userId field is `userId`, not `user_id` 

(See [Models](https://next-auth.js.org/adapters/models))

Unless you changed something in your prisma schema, those are the default values / field names.

**Comment by sergeyshaykhullin** [comment]
@ndom91 Yes, i changed using naming conventions, it works with google provider https://next-auth.js.org/adapters/prisma#naming-conventions

**Comment by ndom91** [comment]
> @ndom91 Yes, i changed using naming conventions, it works with google provider https://next-auth.js.org/adapters/prisma#naming-conventions
> 
> I am using `@map("user_id")` etc.

Okay gotcha, makes sense. 

So `user_id` should actually be there. However `email` still isn't on the account.

File: packages/next-auth/src/providers/vk.ts [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+      token_endpoint_auth_method: "client_secret_post",
+    },
+    token: `https://oauth.vk.com/access_token?v=${apiVersion}`,
+    userinfo: `https://api.vk.com/method/users.get?fields=photo_100&v=${apiVersion}`,
+    profile(result: P) {
+      const profile = result.response?.[0] ?? {}
+      return {
+        id: profile.id,
+        name: [profile.first_name, profile.last_name].filter(Boolean).join(" "),
+        email: null,
```
Comment (sergeyshaykhullin) [inline packages/next-auth/src/providers/vk.ts:279]: @ndom91 Maybe email: null is trying to save accounts.email column?

**Comment by MiXeR54** [comment]
Any updates here? VK provider still not working properly

File: packages/next-auth/src/providers/vk.ts (hunk 3 above)
Comment (ndom91) [reply packages/next-auth/src/providers/vk.ts:279]: No so I meant that the `email` field is not on the `Account` table at all, but the `User` table.

**Comment by pokatomnik** [comment]
I waited for this merge for tvo years.

**Comment by indicozy** [comment]
I also got `invalid_client (client_id is undefined)` error.

File: packages/next-auth/src/providers/vk.ts (hunk 3 above)
Comment (ThangHuuVu) [reply packages/next-auth/src/providers/vk.ts:279]: @ndom91 So if I understand correctly, the problem is that we're passing the `account` object which can contain all kinds of properties (it's extending from `Record<string, unknown>` and  `Partial<TokenSet>` as seen [here](https://github.com/nextauthjs/next-auth/blob/e498483b23273d1bfc81be68339607f88d411bd6/packages/next-auth/src/core/types.ts#L249)) into the adapter function `linkAccount` which expects only the object of type `Account`.

In this case, I think `Prisma` is working as expected. The proper solution IMO is that we should use the object with the desired type `Account` in the `linkAccount` implementation. Does that make sense?

Comment (ndom91) [reply packages/next-auth/src/providers/vk.ts:279]: Yeah that makes sense. So basically drop `email`?

Comment (ThangHuuVu) [reply packages/next-auth/src/providers/vk.ts:279]: Let us drop anything but the expected properties... so I worry that another unexpected property could break the adapter again when used with another provider. WDYT?

Comment (ndom91) [reply packages/next-auth/src/providers/vk.ts:279]: Yeah that makes a lot of sense 👍

**Review by mshd — APPROVED** [review]

**Comment by ndom91** [comment]
@alexTayanovsky can you give me `write` rights to your fork? I'm trying to merge the latest `main` into your branch as it is fairly old.

**Comment by balazsorban44** [comment]
@alexTayanovsky there is a merge conflict, if you could please address that, and allow maintainers to edit, we could get this through the finish line!

**Comment by alexTayanovsky** [comment]
> @alexTayanovsky there is a merge conflict, if you could please address that, and allow maintainers to edit, we could get this through the finish line!
> 
> Otherwise I'm going to close this PR, so someone else can open a new one including the fixes here 🙏 💚

@balazsorban44 done ✅

File: apps/dev/pages/api/auth/[...nextauth].ts [hunk 4]
```diff
@@ -4,7 +4,7 @@ import GitHubProvider from "next-auth/providers/github"
 import Auth0Provider from "next-auth/providers/auth0"
 import KeycloakProvider from "next-auth/providers/keycloak"
 import TwitterProvider, {
-  TwitterLegacy as TwitterLegacyProvider,
+  // TwitterLegacy as TwitterLegacyProvider,
```
Comment (mshd) [inline apps/dev/pages/api/auth/[...nextauth].ts:7]: why is this commented out?

Comment (alexTayanovsky) [reply apps/dev/pages/api/auth/[...nextauth].ts:7]: this is an unused import that caused eslint warning, I guess this line can we removed instead

**Review by balazsorban44 — APPROVED** [review]
Thanks!

**Comment by gosvoh** [comment]
> > @balazsorban44 I tested this fix, it is working, but was unable to save authorized user inside db
> > 
> > Log Output
> > ```
> > [next-auth][error][OAUTH_CALLBACK_HANDLER_ERROR]
> > https://next-auth.js.org/errors#oauth_callback_handler_error
> > Invalid `p.account.create()` invocation in
> > C:\REDACTED\app\node_modules\@next-auth\prisma-adapter\dist\index.js:19:42
> > 
> >   16 },
> >   17 updateUser: ({ id, ...data }) => p.user.update({ where: { id }, data }),
> >   18 deleteUser: (id) => p.user.delete({ where: { id } }),
> > → 19 linkAccount: (data) => p.account.create({
> >        data: {
> >          provider: 'vk',
> >          type: 'oauth',
> >          providerAccountId: 'REDACTED',
> >          access_token: 'REDACTED',
> >          expires_at: 1651408254,
> >          user_id: REDACTED,
> >          ~~~~~~~
> >          email: 'REDACTED',
> >          ~~~~~
> >          userId: 'REDACTED'
> >        }
> >      })
> > 
> > Unknown arg `user_id` in data.user_id for type AccountUncheckedCreateInput. Did you mean `userId`? Available args:
> > type AccountUncheckedCreateInput {
> >   id?: String
> >   userId: String
> >   type: String
> >   provider: String
> >   providerAccountId: String
> >   refresh_token?: String | Null
> >   access_token?: String | Null
> >   expires_at?: Int | Null
> >   token_type?: String | Null
> >   scope?: String | Null
> >   id_token?: String | Null
> >   session_state?: String | Null
> >   oauth_token_secret?: String | Null
> >   oauth_token?: String | Null
> > }
> > Unknown arg `email` in data.email for type AccountUncheckedCreateInput. Did you mean `id`? Available args:
> > type AccountUncheckedCreateInput {
> >   id?: String
> >   userId: String
> >   type: String
> >   provider: String
> >   providerAccountId: String
> >   refresh_token?: String | Null
> >   access_token?: String | Null
> >   expires_at?: Int | Null
> >   token_type?: String | Null
> >   scope?: String | Null
> >   id_token?: String | Null
> >   session_state?: String | Null
> >   oauth_token_secret?: String | Null
> >   oauth_token?: String | Null
> > }
> > 
> >  Error:
> > Invalid `p.account.create()` invocation in
> > C:\REDACTED\app\node_modules\@next-auth\prisma-adapter\dist\index.js:19:42
> > 
> >   16 },
> >   17 updateUser: ({ id, ...data }) => p.user.update({ where: { id }, data }),
> >   18 deleteUser: (id) => p.user.delete({ where: { id } }),
> > → 19 linkAccount: (data) => p.account.create({
> >        data: {
> >          provider: 'vk',
> >          type: 'oauth',
> >          providerAccountId: 'REDACTED',
> >          access_token: 'REDACTED',
> >          expires_at: 1651408254,
> >          user_id: REDACTED,
> >          ~~~~~~~
> >          email: 'REDACTED',
> >          ~~~~~
> >          userId: 'REDACTED'
> >        }
> >      })
> > 
> > Unknown arg `user_id` in data.user_id for type AccountUncheckedCreateInput. Did you mean `userId`? Available args:
> > type AccountUncheckedCreateInput {
> >   id?: String
> >   userId: String
> >   type: String
> >   provider: String
> >   providerAccountId: String
> >   refresh_token?: String | Null
> >   access_token?: String | Null
> >   expires_at?: Int | Null
> >   token_type?: String | Null
> >   scope?: String | Null
> >   id_token?: String | Null
> >   session_state?: String | Null
> >   oauth_token_secret?: String | Null
> >   oauth_token?: String | Null
> > }
> > Unknown arg `email` in data.email for type AccountUncheckedCreateInput. Did you mean `id`? Available args:
> > type AccountUncheckedCreateInput {
> >   id?: String
> >   userId: String
> >   type: String
> >   provider: String
> >   providerAccountId: String
> >   refresh_token?: String | Null
> >   access_token?: String | Null
> >   expires_at?: Int | Null
> >   token_type?: String | Null
> >   scope?: String | Null
> >   id_token?: String | Null
> >   session_state?: String | Null
> >   oauth_token_secret?: String | Null
> >   oauth_token?: String | Null
> > }
> > 
> > 
> >     at Object.validate (C:\REDACTED\app\node_modules\@prisma\client\runtime\index.js:44425:20)
> >     at PrismaClient._executeRequest (C:\REDACTED\app\node_modules\@prisma\client\runtime\index.js:46509:17)
> >     at consumer (C:\REDACTED\app\node_modules\@prisma\client\runtime\index.js:46453:23)
> >     at C:\REDACTED\app\node_modules\@prisma\client\runtime\index.js:46457:76
> >     at runInChildSpan (C:\REDACTED\app\node_modules\@prisma\client\runtime\index.js:45702:12)
> >     at C:\REDACTED\app\node_modules\@prisma\client\runtime\index.js:46457:20
> >     at AsyncResource.runInAsyncScope (node:async_hooks:201:9)
> >     at PrismaClient._request (C:\REDACTED\app\node_modules\@prisma\client\runtime\index.js:46456:86)
> >     at C:\REDACTED\app\node_modules\@prisma\client\runtime\index.js:43088:25
> >     at _callback (C:\REDACTED\app\node_modules\@prisma\client\runtime\index.js:42852:52) {
> >   name: 'LinkAccountError',
> >   code: undefined
> > }
> > ```
> 
> So theres two errors here according to your output:
> 
> 1. `account` model / table doesnt have a field called `email`
> 2. The userId field is `userId`, not `user_id`
> 
> (See [Models](https://next-auth.js.org/adapters/models))
> 
> Unless you changed something in your prisma schema, those are the default values / field names.

The error still appears. In addition to VK, my project uses authorization with Google, Yandex and Mail.ru and they work fine.

**Comment by Almasx** [comment]
Is there any workaround?

**Comment by theneuetimes** [comment]
@Almasx I managed to fix the issue using a quick and dirty work-around - open `@next-auth/prisma-adapter/dist/index.js` and edit line 19 from
```js
linkAccount: (data) => p.account.create({ data }),
```

to

```js
linkAccount: (data) => {
  if(data.provider == 'vk') {
    delete data.user_id;
    delete data.email;
  }

  return p.account.create({ data });
},
```

This is by no means a good or sustainable solution but if you really need VK to work this will do the trick while we wait for an official solution.

**Pull request merged**